internal/controllers/utils/conversion.go

```diff
@@ -0,0 +1,139 @@
+package utils
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	extv1beta1 "k8s.io/api/extensions/v1beta1"
+	netv1 "k8s.io/api/networking/v1"
+	netv1beta1 "k8s.io/api/networking/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kongv1beta1 "github.com/kong/kubernetes-ingress-controller/v2/pkg/apis/configuration/v1beta1"
+)
+
+// UpdateLoadBalancerIngress updates any supported Ingress object with new []netv1.IngressLoadBalancerIngress
+// in a backward-compatible fashion if needed. Update does not happen in case there are no changes detected.
+func UpdateLoadBalancerIngress(
+	ingress client.Object,
+	newAddresses []netv1.IngressLoadBalancerIngress,
+) (updateNeeded bool, err error) {
+	// Convert to netv1 so that we can compare it with newAddresses.
+	oldAddresses, err := ingressToNetV1LoadBalancerIngressStatus(ingress)
+	if err != nil {
+		return false, fmt.Errorf("failed to convert ingress to netv1.Ingress: %w", err)
+	}
+
+	updateNeeded = len(oldAddresses) != len(newAddresses) || !reflect.DeepEqual(oldAddresses, newAddresses)
+	if !updateNeeded {
+		return false, nil
+	}
+
+	switch obj := ingress.(type) {
+	case *netv1.Ingress:
+		obj.Status.LoadBalancer.Ingress = newAddresses
+	case *kongv1beta1.TCPIngress:
+		obj.Status.LoadBalancer.Ingress = netV1ToCoreV1LoadBalancerIngress(newAddresses)
+	case *kongv1beta1.UDPIngress:
+		obj.Status.LoadBalancer.Ingress = netV1ToCoreV1LoadBalancerIngress(newAddresses)
+	case *netv1beta1.Ingress:
+		obj.Status.LoadBalancer.Ingress = netV1ToV1beta1LoadBalancerIngress(newAddresses)
+	case *extv1beta1.Ingress:
+		obj.Status.LoadBalancer.Ingress = netV1ToExtV1beta1LoadBalancerIngress(newAddresses)
+	default:
+		return false, errors.Errorf("unsupported ingress type: %T", obj)
```

Comment:
Can we use `errors` or if you want to format the error with error string then `fmt.Errorf`? `github.com/pkg/errors` is already in read-only mode and is not meant to be supported in the long run

Reply:
Good catch, let's prevent it in the future: 5d19561